packages/api-key/src/services/api-key-module-service.ts

```
@@ -103,36 +142,257 @@ export default class ApiKeyModuleService<TEntity extends ApiKey = ApiKey>
+      updateRequest,
+      sharedContext
+    )
+
+    return revokedApiKeys
   }
 
+  // TODO: Implement
   @InjectTransactionManager("baseRepository_")
   authenticate(
     id: string,
     @MedusaContext() sharedContext: Context = {}
   ): Promise<boolean> {
     return Promise.resolve(false)
   }
+
+  protected async validateCreateApiKeys(
+    data: ApiKeyTypes.CreateApiKeyDTO[],
+    sharedContext: Context = {}
+  ): Promise<void> {
+    if (!data.length) {
+      return
+    }
+
+    // There can only be 2 secret keys at most, and one has to be with a revoked_at date set, so only 1 can be newly created.
+    const secretKeysToCreate = data.filter((k) => k.type === ApiKeyType.SECRET)
+    if (secretKeysToCreate.length > 1) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `You can only create one secret key at a time. You tried to create ${secretKeysToCreate.length} secret keys.`
+      )
+    }
+
+    // There already is a key that is not set to expire/or it hasn't expired
+    const dbSecretKeys = await this.apiKeyService_.list(
+      {
+        type: ApiKeyType.SECRET,
+        revoked_at: null,
+      },
+      {},
+      sharedContext
+    )
+
+    if (dbSecretKeys.length) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `You can only have one active secret key a time. Revoke or delete your existing key before creating a new one.`
+      )
+    }
+  }
+
+  protected async validateRevokeApiKeys(
+    data: ApiKeyTypes.RevokeApiKeyDTO[],
+    sharedContext: Context = {}
+  ): Promise<void> {
+    if (!data.length) {
+      return
+    }
+
+    if (data.some((k) => !k.revoked_by)) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `You must provide a revoked_by field when revoking a key.`
+      )
+    }
+
+    const revokedApiKeys = await this.apiKeyService_.list(
+      {
+        id: data.map((k) => k.id),
```

suggestion: I would extract that map into a const and filter(Boolean), then check if there is a length before ding the list. If the user do not provide the ids, we will list all secret keys

---

@adrien2p wait what? Shouldn't an empty array of ids result in 0 results? Let me write a quick test for it as well.

---

Ok I confirmed, an empty array will not affect anything. It's fine to do what you said as an optimization however.

---

Ok it will actually fail, so what we need to do is add validation and throw an error for invalid requests.